pkg/k2/constructs/construct_evaluator.go

```diff
@@ -845,81 +415,176 @@ func (ce *ConstructEvaluator) evaluateResources(o ResourceOwner, interpolationCt
 	return nil
 }
 
-func GetPropertyFunc(ps *PropertySource) func(string) any {
+func GetPropertyFunc(ps *template2.PropertySource, path string) func(string) any {
 	return func(key string) any {
-		i, ok := ps.GetProperty(fmt.Sprintf("inputs.%s", key))
+		i, ok := ps.GetProperty(fmt.Sprintf("%s.%s", path, key))
 		if !ok {
 			return nil
 		}
 		return i
 	}
 }
 
-func (ce *ConstructEvaluator) templateFunctions(ps *PropertySource) template.FuncMap {
-	funcs := template.FuncMap{}
-	funcs["inputs"] = GetPropertyFunc(ps)
-	return funcs
+func (ce *ConstructEvaluator) evaluateForEachRule(dv *DynamicValueData, rule template2.InputRuleTemplate) error {
+	parentPrefix := dv.resourceKeyPrefix
+
+	ctx := DynamicValueContext{
+		constructs: ce.Constructs,
```

Comment:
Ahh, I see why you changed it to a pointer, but this could just as easily be `&ce.Constructs`